<!DOCTYPE html>

<html>

<head>

<meta charset="utf-8" />
<meta name="generator" content="pandoc" />
<meta http-equiv="X-UA-Compatible" content="IE=EDGE" />

<meta name="viewport" content="width=device-width, initial-scale=1" />

<meta name="author" content="Jaihee Choi" />

<meta name="date" content="2022-10-31" />

<title>SIMICO Vignette</title>

<script>// Pandoc 2.9 adds attributes on both header and div. We remove the former (to
// be compatible with the behavior of Pandoc < 2.8).
document.addEventListener('DOMContentLoaded', function(e) {
  var hs = document.querySelectorAll("div.section[class*='level'] > :first-child");
  var i, h, a;
  for (i = 0; i < hs.length; i++) {
    h = hs[i];
    if (!/^h[1-6]$/i.test(h.tagName)) continue;  // it should be a header h1-h6
    a = h.attributes;
    while (a.length > 0) h.removeAttribute(a[0].name);
  }
});
</script>

<style type="text/css">
  code{white-space: pre-wrap;}
  span.smallcaps{font-variant: small-caps;}
  span.underline{text-decoration: underline;}
  div.column{display: inline-block; vertical-align: top; width: 50%;}
  div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
  ul.task-list{list-style: none;}
    </style>


<style type="text/css">
  code {
    white-space: pre;
  }
  .sourceCode {
    overflow: visible;
  }
</style>
<style type="text/css" data-origin="pandoc">
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { display: inline-block; line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
    color: #aaaaaa;
  }
pre.numberSource { margin-left: 3em; border-left: 1px solid #aaaaaa;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
code span.al { color: #ff0000; font-weight: bold; } /* Alert */
code span.an { color: #60a0b0; font-weight: bold; font-style: italic; } /* Annotation */
code span.at { color: #7d9029; } /* Attribute */
code span.bn { color: #40a070; } /* BaseN */
code span.bu { } /* BuiltIn */
code span.cf { color: #007020; font-weight: bold; } /* ControlFlow */
code span.ch { color: #4070a0; } /* Char */
code span.cn { color: #880000; } /* Constant */
code span.co { color: #60a0b0; font-style: italic; } /* Comment */
code span.cv { color: #60a0b0; font-weight: bold; font-style: italic; } /* CommentVar */
code span.do { color: #ba2121; font-style: italic; } /* Documentation */
code span.dt { color: #902000; } /* DataType */
code span.dv { color: #40a070; } /* DecVal */
code span.er { color: #ff0000; font-weight: bold; } /* Error */
code span.ex { } /* Extension */
code span.fl { color: #40a070; } /* Float */
code span.fu { color: #06287e; } /* Function */
code span.im { } /* Import */
code span.in { color: #60a0b0; font-weight: bold; font-style: italic; } /* Information */
code span.kw { color: #007020; font-weight: bold; } /* Keyword */
code span.op { color: #666666; } /* Operator */
code span.ot { color: #007020; } /* Other */
code span.pp { color: #bc7a00; } /* Preprocessor */
code span.sc { color: #4070a0; } /* SpecialChar */
code span.ss { color: #bb6688; } /* SpecialString */
code span.st { color: #4070a0; } /* String */
code span.va { color: #19177c; } /* Variable */
code span.vs { color: #4070a0; } /* VerbatimString */
code span.wa { color: #60a0b0; font-weight: bold; font-style: italic; } /* Warning */

</style>
<script>
// apply pandoc div.sourceCode style to pre.sourceCode instead
(function() {
  var sheets = document.styleSheets;
  for (var i = 0; i < sheets.length; i++) {
    if (sheets[i].ownerNode.dataset["origin"] !== "pandoc") continue;
    try { var rules = sheets[i].cssRules; } catch (e) { continue; }
    for (var j = 0; j < rules.length; j++) {
      var rule = rules[j];
      // check if there is a div.sourceCode rule
      if (rule.type !== rule.STYLE_RULE || rule.selectorText !== "div.sourceCode") continue;
      var style = rule.style.cssText;
      // check if color or background-color is set
      if (rule.style.color === '' && rule.style.backgroundColor === '') continue;
      // replace div.sourceCode by a pre.sourceCode rule
      sheets[i].deleteRule(j);
      sheets[i].insertRule('pre.sourceCode{' + style + '}', j);
    }
  }
})();
</script>




<style type="text/css">body {
background-color: #fff;
margin: 1em auto;
max-width: 700px;
overflow: visible;
padding-left: 2em;
padding-right: 2em;
font-family: "Open Sans", "Helvetica Neue", Helvetica, Arial, sans-serif;
font-size: 14px;
line-height: 1.35;
}
#TOC {
clear: both;
margin: 0 0 10px 10px;
padding: 4px;
width: 400px;
border: 1px solid #CCCCCC;
border-radius: 5px;
background-color: #f6f6f6;
font-size: 13px;
line-height: 1.3;
}
#TOC .toctitle {
font-weight: bold;
font-size: 15px;
margin-left: 5px;
}
#TOC ul {
padding-left: 40px;
margin-left: -1.5em;
margin-top: 5px;
margin-bottom: 5px;
}
#TOC ul ul {
margin-left: -2em;
}
#TOC li {
line-height: 16px;
}
table {
margin: 1em auto;
border-width: 1px;
border-color: #DDDDDD;
border-style: outset;
border-collapse: collapse;
}
table th {
border-width: 2px;
padding: 5px;
border-style: inset;
}
table td {
border-width: 1px;
border-style: inset;
line-height: 18px;
padding: 5px 5px;
}
table, table th, table td {
border-left-style: none;
border-right-style: none;
}
table thead, table tr.even {
background-color: #f7f7f7;
}
p {
margin: 0.5em 0;
}
blockquote {
background-color: #f6f6f6;
padding: 0.25em 0.75em;
}
hr {
border-style: solid;
border: none;
border-top: 1px solid #777;
margin: 28px 0;
}
dl {
margin-left: 0;
}
dl dd {
margin-bottom: 13px;
margin-left: 13px;
}
dl dt {
font-weight: bold;
}
ul {
margin-top: 0;
}
ul li {
list-style: circle outside;
}
ul ul {
margin-bottom: 0;
}
pre, code {
background-color: #f7f7f7;
border-radius: 3px;
color: #333;
white-space: pre-wrap; 
}
pre {
border-radius: 3px;
margin: 5px 0px 10px 0px;
padding: 10px;
}
pre:not([class]) {
background-color: #f7f7f7;
}
code {
font-family: Consolas, Monaco, 'Courier New', monospace;
font-size: 85%;
}
p > code, li > code {
padding: 2px 0px;
}
div.figure {
text-align: center;
}
img {
background-color: #FFFFFF;
padding: 2px;
border: 1px solid #DDDDDD;
border-radius: 3px;
border: 1px solid #CCCCCC;
margin: 0 5px;
}
h1 {
margin-top: 0;
font-size: 35px;
line-height: 40px;
}
h2 {
border-bottom: 4px solid #f7f7f7;
padding-top: 10px;
padding-bottom: 2px;
font-size: 145%;
}
h3 {
border-bottom: 2px solid #f7f7f7;
padding-top: 10px;
font-size: 120%;
}
h4 {
border-bottom: 1px solid #f7f7f7;
margin-left: 8px;
font-size: 105%;
}
h5, h6 {
border-bottom: 1px solid #ccc;
font-size: 105%;
}
a {
color: #0033dd;
text-decoration: none;
}
a:hover {
color: #6666ff; }
a:visited {
color: #800080; }
a:visited:hover {
color: #BB00BB; }
a[href^="http:"] {
text-decoration: underline; }
a[href^="https:"] {
text-decoration: underline; }

code > span.kw { color: #555; font-weight: bold; } 
code > span.dt { color: #902000; } 
code > span.dv { color: #40a070; } 
code > span.bn { color: #d14; } 
code > span.fl { color: #d14; } 
code > span.ch { color: #d14; } 
code > span.st { color: #d14; } 
code > span.co { color: #888888; font-style: italic; } 
code > span.ot { color: #007020; } 
code > span.al { color: #ff0000; font-weight: bold; } 
code > span.fu { color: #900; font-weight: bold; } 
code > span.er { color: #a61717; background-color: #e3d2d2; } 
</style>




</head>

<body>




<h1 class="title toc-ignore">SIMICO Vignette</h1>
<h4 class="author">Jaihee Choi</h4>
<h4 class="date">2022-10-31</h4>



<div id="introduction" class="section level3">
<h3>Introduction</h3>
<p>This package implements the multiple outcome test and multiple burden
test specified in “Variance-components Tests for Genetic Association
with Multiple Interval-censored Outcomes”.</p>
<p>Users can test a set of genetic variants with multiple
interval-censored outcomes. Interval-censored outcomes arises when the
exact time of onset of an event of interest is unknown but is known to
fall between two time points. This type of data occurs in many
large-scale genetic compendiums, such as the UK Biobank. Due to the
large scale nature of these compendiums, one way that information is
collected is via periodic questionnaires, causing much of the data to be
interval-censored.</p>
<p>The steps to implement the functions, the specified form of the data
inputs, and a worked example is shown below.</p>
</div>
<div id="data-form" class="section level3">
<h3>Data form</h3>
<p>The data inputs for the SIMICO functions follows a specific form. All
the left and right design matrices and interval times must be put in a
list item indexed by the outcome number. Some code that achieves this
data format is:</p>
<p>For the left and right design matrices (these are the matrices
containing the covariate information with the cubic spline terms. The
cubic spline terms can be generated by calling
ICSKAT::make_IC_dmat())</p>
<pre><code>dmats &lt;- list(right_dmat = [right design matrix], left_dmat = [left design matrix])
allData &lt;- list(xDats = list(dmats))</code></pre>
<pre><code>length(xDats)</code></pre>
<p>should equal the number of outcomes</p>
<pre><code>attributes(xAll$xDats[[1]]$dmats)}</code></pre>
<p>should be “right_dmat” “left_dmat”.</p>
<p>Additionally, the data input list should also contain the left and
right censoring information:<br />
tpos_all and obs_all are both n x k matrices containing binary 0/1 for
whether the individual is left or right censored for that outcome,
respectively. If only the matrices of the left and right time (named
leftTimesMat and rightTimesMat in the below example) are available, then
one can convert these values to censoring terms by the following, for k
outcomes:</p>
<pre><code>obs_all &lt;- matrix(NA, nrow= n, ncol = k)
tpos_all &lt;- matrix(NA, nrow= n, ncol = k)

# Loop through all the outcomes
for(i in 1:k){
    obs_all[, i] &lt;- ifelse(rightTimesMat[, i] == Inf, 0, 1)
    tpos_all[, i] &lt;- ifelse(leftTimesMat[, i] == 0, 0, 1)
}</code></pre>
<p>All the data should be together in one list with the following
form:</p>
<pre><code>allData &lt;- list(xDats = xDats, ts_all = tpos_all, ob_all = obs_all)</code></pre>
<p>Accessing the right design matrix for ONE outcome should be:<br />
</p>
<pre><code>allData$xDats[[OUTCOME NUMBER]]$dmats$right_dmat</code></pre>
<p>Accessing the n x 1 vector of whether the observations were
left-censored or not should be:<br />
</p>
<pre><code>allData$obs_all[,OUTCOME NUMBER]</code></pre>
</div>
<div id="instructions-for-simico-functions" class="section level2">
<h2>Instructions for SIMICO Functions</h2>
<p>There are three steps to running the multiple outcome test with this
package:</p>
<div id="step-1" class="section level3">
<h3>Step 1:</h3>
<p>Generate the left and right time intervals , as well as the left and
right design matrices for the covariates. This can all be done through
the simico_gen_dat() function, where the inputs are:<br />
- bhFunInv is the inverse of the baseline hazard function.<br />
- obsTimes is the vector of the intended observation times.<br />
- windowHalf is the amount of time before or after the intended obsTimes
that a visit might take place.<br />
- n is the total number of observations.<br />
- p is the total number of covariates.<br />
- k is the total number of outcomes.<br />
- tausq is the variance of the subject specific random effect.<br />
- gMatCausal is the matrix of subsetted genetic information for only a
select causal SNPs.<br />
- xMat is the matrix of covariates.<br />
- effectSizes is the vector of genetic effect sizes. Should be entered
as a vector the same length as the number of outcomes.<br />
</p>
</div>
<div id="step-2" class="section level3">
<h3>Step 2:</h3>
<p>Fit the null model. You only need to do this once for all SNP-sets to
be tested. The call is simico_fit_null(), and you need the following
arguments:<br />
-init_beta is a vector holding the initial guess at the covariates for
all of the outcomes plus the variance of the subject specific random
effect. The number of elements should be equal to the number of columns
in the design matrices times the number of outcomes plus 1 (for the
subject specific random effect). Usually this can be initialized to a
vector of 0s or 1s. If you happen to have a good idea of what the
coefficients are, then this will speed up convergence.<br />
- epsilon is the stopping criterion for NR.<br />
- xDats is the list of left and right design matrices. The form is the
same as specified earlier.<br />
- lt_all is the n x k matrix of left times.<br />
- rt_all is the n x k matrix of right times.<br />
- k is the total number of outcomes.<br />
- d is the total number of quadrature nodes.<br />
</p>
</div>
<div id="step-3" class="section level3">
<h3>Step 3</h3>
<p>Call simico_out() for each set of SNPs that you want to test. The G
argument should be the n x q matrix of genotypes. a1 and a2 are the
shape parameters for the Beta distribution, which are used to weight
certain variants over others by their MAF. You will also need xDats,
lt_all, rt_all, k, and d, which have been defined above. Finally, you
will need null_beta and Itt, which come directly from simico_fit_null().
A worked example is below.</p>
<p>Worked Example:</p>
<p>Suppose we are interested in testing whether a specific gene is
associated with two different correlated outcomes: time until a fall and
time until a fracture. We will simulate event times for 5,000 subjects
under a proportional hazards model with baseline cumulative hazard
function H(t)=t. We will set four observations times at times 1, 2, 3,
and 4. Each subject will have a 10% chance of missing any given visit.
The genetic data set will consist of 50 SNPs in the gene of interest. We
will use 100 quadrature nodes. The subject specific random effect <span class="math inline">\(b_i\)</span> will be drawn from a N(0,1)
distribution. The genetic matrices were generated to have mean
parameters uniformly distributed between (0.01, 0.05) and common
pairwise correlation of 0.1. To simulate the data, we specifiy 5 causal
variants with a MAF cutoff of 0.05. We simulated two fixed
covariates–one generated from a N(0,2) and the other from a
Binomaial(0.5). For the genetic effect sizes we chose (.03, .15).</p>
<p>Example with two outcomes</p>
<div class="sourceCode" id="cb8"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb8-1"><a href="#cb8-1" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(SIMICO)</span>
<span id="cb8-2"><a href="#cb8-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb8-3"><a href="#cb8-3" aria-hidden="true" tabindex="-1"></a><span class="co"># Set two outcomes</span></span>
<span id="cb8-4"><a href="#cb8-4" aria-hidden="true" tabindex="-1"></a>k <span class="ot">=</span> <span class="dv">2</span></span>
<span id="cb8-5"><a href="#cb8-5" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb8-6"><a href="#cb8-6" aria-hidden="true" tabindex="-1"></a><span class="co"># Set number of observations</span></span>
<span id="cb8-7"><a href="#cb8-7" aria-hidden="true" tabindex="-1"></a>n <span class="ot">=</span> <span class="dv">5000</span></span>
<span id="cb8-8"><a href="#cb8-8" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb8-9"><a href="#cb8-9" aria-hidden="true" tabindex="-1"></a><span class="co"># Set number of covariates</span></span>
<span id="cb8-10"><a href="#cb8-10" aria-hidden="true" tabindex="-1"></a>p <span class="ot">=</span> <span class="dv">2</span></span>
<span id="cb8-11"><a href="#cb8-11" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb8-12"><a href="#cb8-12" aria-hidden="true" tabindex="-1"></a><span class="co"># Set number of SNPs</span></span>
<span id="cb8-13"><a href="#cb8-13" aria-hidden="true" tabindex="-1"></a>q <span class="ot">=</span> <span class="dv">50</span></span>
<span id="cb8-14"><a href="#cb8-14" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb8-15"><a href="#cb8-15" aria-hidden="true" tabindex="-1"></a><span class="co"># Set number of quadrature nodes</span></span>
<span id="cb8-16"><a href="#cb8-16" aria-hidden="true" tabindex="-1"></a>d <span class="ot">=</span> <span class="dv">100</span></span>
<span id="cb8-17"><a href="#cb8-17" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb8-18"><a href="#cb8-18" aria-hidden="true" tabindex="-1"></a><span class="co"># Variance of subject-specific random effect</span></span>
<span id="cb8-19"><a href="#cb8-19" aria-hidden="true" tabindex="-1"></a>tauSq <span class="ot">=</span> <span class="dv">1</span></span>
<span id="cb8-20"><a href="#cb8-20" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb8-21"><a href="#cb8-21" aria-hidden="true" tabindex="-1"></a><span class="co"># Pairwise correlation</span></span>
<span id="cb8-22"><a href="#cb8-22" aria-hidden="true" tabindex="-1"></a>rho <span class="ot">=</span> <span class="fl">0.1</span></span>
<span id="cb8-23"><a href="#cb8-23" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb8-24"><a href="#cb8-24" aria-hidden="true" tabindex="-1"></a><span class="co"># Minor Allele Frequencey Cutoff</span></span>
<span id="cb8-25"><a href="#cb8-25" aria-hidden="true" tabindex="-1"></a>Causal.MAF.Cutoff <span class="ot">=</span> .<span class="dv">05</span></span>
<span id="cb8-26"><a href="#cb8-26" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb8-27"><a href="#cb8-27" aria-hidden="true" tabindex="-1"></a><span class="co"># Total number of causal variants</span></span>
<span id="cb8-28"><a href="#cb8-28" aria-hidden="true" tabindex="-1"></a>num <span class="ot">=</span> <span class="dv">5</span></span>
<span id="cb8-29"><a href="#cb8-29" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb8-30"><a href="#cb8-30" aria-hidden="true" tabindex="-1"></a><span class="co"># Define the effect sizes</span></span>
<span id="cb8-31"><a href="#cb8-31" aria-hidden="true" tabindex="-1"></a>effectSizes <span class="ot">&lt;-</span> <span class="fu">c</span>(.<span class="dv">03</span>, .<span class="dv">15</span>)</span>
<span id="cb8-32"><a href="#cb8-32" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb8-33"><a href="#cb8-33" aria-hidden="true" tabindex="-1"></a><span class="co"># the baseline cumulative hazard function</span></span>
<span id="cb8-34"><a href="#cb8-34" aria-hidden="true" tabindex="-1"></a>bhFunInv <span class="ot">&lt;-</span> <span class="cf">function</span>(x) {x}</span>
<span id="cb8-35"><a href="#cb8-35" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb8-36"><a href="#cb8-36" aria-hidden="true" tabindex="-1"></a><span class="fu">set.seed</span>(<span class="dv">1</span>)</span>
<span id="cb8-37"><a href="#cb8-37" aria-hidden="true" tabindex="-1"></a><span class="co"># Fixed effects</span></span>
<span id="cb8-38"><a href="#cb8-38" aria-hidden="true" tabindex="-1"></a>xMat <span class="ot">&lt;-</span> <span class="fu">cbind</span>(<span class="fu">rnorm</span>(n, <span class="at">mean =</span> <span class="dv">0</span>, <span class="at">sd =</span> <span class="dv">2</span>), <span class="fu">rbinom</span>(<span class="at">n=</span>n, <span class="at">size=</span><span class="dv">1</span>, <span class="at">prob=</span><span class="fl">0.5</span>))</span>
<span id="cb8-39"><a href="#cb8-39" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb8-40"><a href="#cb8-40" aria-hidden="true" tabindex="-1"></a><span class="co"># Genetic effects</span></span>
<span id="cb8-41"><a href="#cb8-41" aria-hidden="true" tabindex="-1"></a>gMat <span class="ot">&lt;-</span> <span class="fu">sim_gmat</span>(n, q, rho)</span>
<span id="cb8-42"><a href="#cb8-42" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb8-43"><a href="#cb8-43" aria-hidden="true" tabindex="-1"></a><span class="co"># Get indices to specific select causal variants</span></span>
<span id="cb8-44"><a href="#cb8-44" aria-hidden="true" tabindex="-1"></a>idx <span class="ot">&lt;-</span> <span class="fu">Get_CausalSNPs_bynum</span>(gMat, num, Causal.MAF.Cutoff)</span>
<span id="cb8-45"><a href="#cb8-45" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb8-46"><a href="#cb8-46" aria-hidden="true" tabindex="-1"></a><span class="co"># Subset the gMat</span></span>
<span id="cb8-47"><a href="#cb8-47" aria-hidden="true" tabindex="-1"></a>gMatCausal <span class="ot">&lt;-</span> gMat[,idx]</span>
<span id="cb8-48"><a href="#cb8-48" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb8-49"><a href="#cb8-49" aria-hidden="true" tabindex="-1"></a><span class="co"># True model has nothing</span></span>
<span id="cb8-50"><a href="#cb8-50" aria-hidden="true" tabindex="-1"></a>fixedMat <span class="ot">&lt;-</span> <span class="fu">matrix</span>(<span class="at">data=</span><span class="dv">0</span>, <span class="at">nrow=</span>n, <span class="at">ncol=</span>k)</span>
<span id="cb8-51"><a href="#cb8-51" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb8-52"><a href="#cb8-52" aria-hidden="true" tabindex="-1"></a><span class="co"># Generate the multiple outcomes</span></span>
<span id="cb8-53"><a href="#cb8-53" aria-hidden="true" tabindex="-1"></a>exampleDat <span class="ot">&lt;-</span> <span class="fu">simico_gen_dat</span>(<span class="at">bhFunInv =</span> bhFunInv, <span class="at">obsTimes =</span> <span class="dv">1</span><span class="sc">:</span><span class="dv">3</span>,</span>
<span id="cb8-54"><a href="#cb8-54" aria-hidden="true" tabindex="-1"></a>                             <span class="at">windowHalf =</span> <span class="fl">0.1</span>, <span class="at">n =</span> n, <span class="at">p =</span> p, <span class="at">k =</span> k, </span>
<span id="cb8-55"><a href="#cb8-55" aria-hidden="true" tabindex="-1"></a>                             <span class="at">tauSq =</span> tauSq, <span class="at">gMatCausal =</span> gMatCausal,</span>
<span id="cb8-56"><a href="#cb8-56" aria-hidden="true" tabindex="-1"></a>                             <span class="at">xMat =</span> xMat, <span class="at">effectSizes =</span> effectSizes)</span>
<span id="cb8-57"><a href="#cb8-57" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb8-58"><a href="#cb8-58" aria-hidden="true" tabindex="-1"></a><span class="co"># Set the initial estimate values</span></span>
<span id="cb8-59"><a href="#cb8-59" aria-hidden="true" tabindex="-1"></a>init_beta <span class="ot">&lt;-</span><span class="fu">c</span> (<span class="fu">rep</span>(<span class="fu">c</span>(<span class="dv">0</span>, <span class="dv">0</span>, <span class="dv">0</span>, <span class="dv">1</span>, <span class="dv">0</span>), k), <span class="dv">1</span>)</span>
<span id="cb8-60"><a href="#cb8-60" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb8-61"><a href="#cb8-61" aria-hidden="true" tabindex="-1"></a><span class="co"># Run the newton-raphson</span></span>
<span id="cb8-62"><a href="#cb8-62" aria-hidden="true" tabindex="-1"></a>nullFit <span class="ot">&lt;-</span> <span class="fu">simico_fit_null</span>(<span class="at">init_beta =</span> init_beta, <span class="at">epsilon =</span> <span class="dv">10</span><span class="sc">^-</span><span class="dv">5</span>, </span>
<span id="cb8-63"><a href="#cb8-63" aria-hidden="true" tabindex="-1"></a>                           <span class="at">xDats =</span> exampleDat<span class="sc">$</span>fullDat<span class="sc">$</span>xDats, </span>
<span id="cb8-64"><a href="#cb8-64" aria-hidden="true" tabindex="-1"></a>                           <span class="at">lt_all =</span> exampleDat<span class="sc">$</span>leftTimesMat, <span class="at">rt_all =</span> exampleDat<span class="sc">$</span>rightTimesMat, </span>
<span id="cb8-65"><a href="#cb8-65" aria-hidden="true" tabindex="-1"></a>                           <span class="at">k =</span> k, <span class="at">d =</span> d)</span></code></pre></div>
<pre><code>## [1] 2
##           [,1]
## [1,] 0.5908218
## [1] 3
##            [,1]
## [1,] 0.00100551
## [1] 4
##              [,1]
## [1,] 1.128265e-06</code></pre>
<div class="sourceCode" id="cb10"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb10-1"><a href="#cb10-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Get the test statistics p-values</span></span>
<span id="cb10-2"><a href="#cb10-2" aria-hidden="true" tabindex="-1"></a>out <span class="ot">&lt;-</span> <span class="fu">simico_out</span>(<span class="at">nullFit =</span> nullFit<span class="sc">$</span>beta_fit, <span class="at">xDats =</span> exampleDat<span class="sc">$</span>fullDat<span class="sc">$</span>xDats, </span>
<span id="cb10-3"><a href="#cb10-3" aria-hidden="true" tabindex="-1"></a>                  <span class="at">lt_all =</span> exampleDat<span class="sc">$</span>leftTimesMat, <span class="at">rt_all =</span> exampleDat<span class="sc">$</span>rightTimesMat, </span>
<span id="cb10-4"><a href="#cb10-4" aria-hidden="true" tabindex="-1"></a>                  <span class="at">Itt =</span> nullFit<span class="sc">$</span>jmat, <span class="at">a1 =</span> <span class="dv">1</span>, <span class="at">a2 =</span> <span class="dv">25</span>, </span>
<span id="cb10-5"><a href="#cb10-5" aria-hidden="true" tabindex="-1"></a>                  <span class="at">G =</span> gMat, <span class="at">k  =</span> k, <span class="at">d =</span> d)</span>
<span id="cb10-6"><a href="#cb10-6" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb10-7"><a href="#cb10-7" aria-hidden="true" tabindex="-1"></a><span class="co">#  results</span></span>
<span id="cb10-8"><a href="#cb10-8" aria-hidden="true" tabindex="-1"></a><span class="co"># Score statistic</span></span>
<span id="cb10-9"><a href="#cb10-9" aria-hidden="true" tabindex="-1"></a>(out<span class="sc">$</span>multQ)</span></code></pre></div>
<pre><code>##         [,1]
## [1,] 3641276</code></pre>
<div class="sourceCode" id="cb12"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb12-1"><a href="#cb12-1" aria-hidden="true" tabindex="-1"></a><span class="co"># P-values</span></span>
<span id="cb12-2"><a href="#cb12-2" aria-hidden="true" tabindex="-1"></a>(out<span class="sc">$</span>multP)</span></code></pre></div>
<pre><code>## [1] 0.0004766614</code></pre>
</div>
</div>



<!-- code folding -->


<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    script.src  = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML";
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>

</body>
</html>
